Remove FileInput inheritance from Psr and HttpServer implementations

ramchale opened the pull request:
|    Q          |   A
|-------------- | ------
| Documentation | no
| Bugfix        | no
| BC Break      | yes
| New Feature   | no
| RFC           | yes
| QA            | yes

### Description

The inheritance was only used to allow access to protected properties of FileInput. This refactor removes the need for this and makes the flow of logic slightly easier to follow.

ramchale commented:
The intent is to make this a little easier to understand and maintain going forward. If people are happy with the approach then the tests in PsrFileInputDecoratorTest and HttpServerFileInputDecoratorTest should mostly move into a FileInputTest file and be de-duplicated as much as possible. We could also consider renaming PsrFileInputDecorator and HttpServerFileInputDecorator to be clearer on their purpose

gsteel commented on src/FileInput/HttpServerFileInputDecorator.php:
```diff
@@ -29,15 +29,10 @@
  * @final
  */
-class HttpServerFileInputDecorator extends FileInput implements FileInputDecoratorInterface
+class HttpServerFileInputDecorator implements FileInputDecoratorInterface
```
Can we use the `final` keyword and make this `@internal`?

FYI, instead of `@internal` you can

```php
/**
 * @psalm-internal Laminas\InputFilter
 * @psalm-internal LaminasTest\InputFilter
 */
```

… to minimise Psalm issues

gsteel commented on src/FileInput/FileInputDecoratorInterface.php:
```diff
@@ -13,22 +16,11 @@
  */
 interface FileInputDecoratorInterface
```
`@internal`?

gsteel commented on src/FileInput.php:
```diff
@@ -25,13 +25,13 @@
 
     /** @var bool */
-    protected $autoPrependUploadValidator = true;
+    private $autoPrependUploadValidator = true;
```
Can this property be made immutable rather than passing it to the decorator constructor by ref? By ref stuff really sucks and is horrible to test - ideally we could drop the getter and setter, or, make the input itself the constructor arg so the decorator can call the getter when it matters.

That said, you're passing the validator chain as a method arg - why not mutate the validator chain inside `FileInput` prior to executing the decorator when necessary, then the decorator doesn't need access to this property at all - i.e. the decorator accepts the validator chain as-is.

ramchale replied:
~~If we remove get/setAutoPrependUploadValidator from FileInput then yep, I can handle autoPrependUploadValidator in a less horrible way 😄~~

~~The initial approach I looked at ran into issues with calls to FileInput->setValidatorChain after construction of the decorator. The injectUploadValidator methods in each decorator are putting in different validators, but I'll give it more of a look, as there's probably a nicer way of handling this~~

Scratch all that, clearly needed to walk away from the laptop for a bit. Given 'fileuploadfile' is an alias for the UploadFile validator this should work fine

gsteel reviewed (commented):
This will be a huge improvement in readability and maintainability 👍 

I'm in favour of renaming the 'Decorators' and explicitly marking all of that stuff as internal implementation detail.

ramchale commented:
@gsteel I think it ready for another look.
Despite raising it, I can't actually think of a better name for the decorators, they're handling some validation and filtering so it'd need to be something that neatly encompasses both.

gsteel commented on test/FileInput/FileInputTest.php:
```diff
@@ -0,0 +1,1146 @@
+        self::assertTrue($this->input->isRequired());
+        $this->input->setValue($raw);
+
+        $uploadMock = $this->createMock(Validator\File\UploadFile::class);
```
Pretty sure this is final in 3.0 so mocking is out of the window. Can you compose a chain without a mock here, or, fix it when we upgrade to validator 3.x.

ramchale replied:
Good point. I've switched it to use values that fail validation to avoid having to fake uploading a file, but it is now using an actual instance of UploadFile

gsteel reviewed (approved):
Thanks @ramchale - Just one note about mocking the upload validator, but otherwise LGTM 👍

gsteel commented:
> Despite raising it, I can't actually think of a better name for the decorators, they're handling some validation and filtering so it'd need to be something that neatly encompasses both.

Maybe go with something generic such as `Helper`. I think it's a good idea to change the name because they really aren't decorators at all.

ramchale committed:
Remove FileInput inheritance from Psr and HttpServer implementations

The inheritance was only used to allow access to protected properties of FileInput.
This refactor removes the need for this and makes the flow of logic slightly easier to follow.

- Remove inheritance of FileInput
- Rename HttpServerFileInputDecorator to HttpServerFileInputHandler
- Rename PsrFileInputDecorator to PsrFileInputHandler
- Reorganise the tests to reduce duplication
- Replace mocks in test of NotEmpty and FileUpload validators with instances for validator v3 compatibility

Signed-off-by: ramchale <legionsofbob@gmail.com>

ramchale commented:
> > Despite raising it, I can't actually think of a better name for the decorators, they're handling some validation and filtering so it'd need to be something that neatly encompasses both.
> 
> Maybe go with something generic such as `Helper`. I think it's a good idea to change the name because they really aren't decorators at all.

👍  I've gone with Handler

ramchale commented:
I've squashed this, so if you're happy with the last couple of bits, I think it's good to merge

gsteel commented on src/FileInput/FileInputDecoratorInterface.php:
```diff
@@ -4,31 +4,26 @@
+ * @psalm-internal Laminas\InputFilter
+ * @psalm-internal LaminasTest\InputFilter
  */
 interface FileInputDecoratorInterface
```
You should also rename this to something like `FileInputHandlerInterface`

gsteel commented on src/FileInput.php:
```diff
@@ -79,7 +77,7 @@ public function getValue()
             return $this->value;
         }
-        return $this->implementation->getValue();
+        return $this->implementation->filterValue($this->value, $this->isValid, $this->getFilterChain());
```
Can we rename this private prop `$handler` instead of `$implementation`?

gsteel commented on src/FileInput.php:
```diff
@@ -172,20 +175,42 @@ private function createDecoratorImplementation($value)
     {
```
And this method should be renamed too.

gsteel commented on src/FileInput/FileInputDecoratorInterface.php:
```diff
@@ -4,31 +4,26 @@
-     */
-    public static function isEmptyFileDecorator($rawValue);
+    /** Checks if the raw input value is an empty file input eg: no file was uploaded */
+    public static function isEmptyFileDecorator(mixed $rawValue): bool;
```
Can this just be `isEmptyFile`

ramchale commented:
@gsteel thanks. Not sure how I missed those 😓 , should be cleaned up now

gsteel reviewed (approved):
Thanks @ramchale 👍
